libs/core/async_sycl/include/hpx/async_sycl/sycl_executor.hpp

```diff
@@ -0,0 +1,182 @@
+//  Copyright (c) 2022 Gregor Daiß
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/config.hpp>
+#include <hpx/async_sycl/sycl_future.hpp>
+#include <hpx/errors/exception.hpp>
+#include <hpx/errors/try_catch_exception_ptr.hpp>
+#include <hpx/execution_base/execution.hpp>
+#include <hpx/execution_base/traits/is_executor.hpp>
+#include <hpx/futures/future.hpp>
+#include <hpx/futures/traits/future_access.hpp>
+
+#include <cstddef>
+#include <exception>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace sycl { namespace experimental {
+
+    struct sycl_executor
+    {
+        using future_type = hpx::future<void>;
+
+        // -------------------------------------------------------------------------
+        /// Create a SYCL executor (based on a sycl queue)
+        explicit sycl_executor(cl::sycl::default_selector selector)
+          : command_queue(selector, cl::sycl::property::queue::in_order{})
+        {
+        }
+
+        // TODO Future work: Add more constructors as needed (in case we need different queue constructors)
+        // -------------------------------------------------------------------------
+        // Queue will be cleaned up by its own destructor
+        ~sycl_executor() {}
+
+        /// Get future for this command_queue (NOTE will be more efficient if
+        /// an event is provided -- otherwise a dummy kernel must be submitted to
+        /// get an event)
+        future_type get_future()
+        {
+            // The SYCL standard does not include a eventRecord method Instead
+            // we have to submit some dummy function and use the event the
+            // launch returns
+            cl::sycl::event event = command_queue.submit(
+                [&](cl::sycl::handler& h) { h.single_task([]() {}); });
+            return detail::get_future(event);
+        }
+
+        /// Get future for that becomes ready when the given event completes
+        inline future_type get_future(cl::sycl::event event)
+        {
+            return detail::get_future(event);
+        }
+
+#if defined(__INTEL_LLVM_COMPILER)
+        // To find the correct overload (or any at all actually) we need to add the
+        // code_location argument which is the last argument in every queue member
+        // function in the intel oneapi sycl implementation.  As far as I can tell
+        // it is usually invisible from the user-side since it is using a default
+        // argument (code_location::current())
+
+        /// sycl::queue::member_function type with code_location parameter
+        template <typename... Params>
+        using queue_function_ptr_t = cl::sycl::event (cl::sycl::queue::*)(
+            Params..., const cl::sycl::detail::code_location&);
```

Comment:
Would you mind using 'east-const' notation, if possible? This would make the overall codebase more uniform.

Reply:
Sure, not a problem! I have changed to east-const in 951382f48d56c1b90d460ed996e57fa3c670cbdd